tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/Refiner.scala

```diff
@@ -95,3 +107,137 @@ final class Refiner(config: CommonPhaseConfig) {
   }
 
 }
+
+private object Refiner {
+  private class InputProvider extends Analyzer.InputProvider {
+    private var linkedClassesByName: Map[String, LinkedClass] = _
+    private val cache = mutable.Map.empty[String, LinkedClassInfoCache]
+
+    def update(linkedClassesByName: Map[String, LinkedClass]): Unit = {
+      this.linkedClassesByName = linkedClassesByName
+    }
+
+    def classesWithEntryPoints(): TraversableOnce[String] = {
+      for {
+        linkedClass <- linkedClassesByName.valuesIterator
+        if linkedClass.hasEntryPoint
+      } yield {
+        linkedClass.encodedName
+      }
+    }
+
+    def loadInfo(encodedName: String): Option[Infos.ClassInfo] =
+      getCache(encodedName).map(_.loadInfo(linkedClassesByName(encodedName)))
+
+    private def getCache(encodedName: String): Option[LinkedClassInfoCache] = {
+      cache.get(encodedName).orElse {
+        if (linkedClassesByName.contains(encodedName)) {
+          val fileCache = new LinkedClassInfoCache
+          cache += encodedName -> fileCache
+          Some(fileCache)
+        } else {
+          None
+        }
+      }
+    }
+
+    def cleanAfterRun(): Unit = {
+      linkedClassesByName = null
+      cache.retain((_, linkedClassCache) => linkedClassCache.cleanAfterRun())
+    }
+  }
+
+  private class LinkedClassInfoCache {
```

Comment:
I think the main reason this is so ugly is that there is no `MethodInfoCache`.

If you add a `MethodInfoCache`, updating the class def becomes:
- Update all MethodInfoCaches
- Regenerate ClassInfo pulling it in from the method caches (don't even bother caching it, it's just copying stuff).

Have you considered that?